# Vikash Patel

## JSON Exercise

Using data in the file 'data/world_bank_projects.json',

 1. Find the 10 countries with most projects
 2. Find the top 10 major project themes (using column 'mjtheme_namecode')
 3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

## 1.) Finding the top 10 countries with most projects:

In [1]:
import pandas as pd

wbp_df = pd.read_json('data/world_bank_projects.json')    #reads in the JSON file as a pandas DataFrame
wbp_df.head()

_id  approvalfy board_approval_month  \
0  {'$oid': '52b213b38594d8a2be17c780'}        1999             November   
1  {'$oid': '52b213b38594d8a2be17c781'}        2015             November   
2  {'$oid': '52b213b38594d8a2be17c782'}        2014             November   
3  {'$oid': '52b213b38594d8a2be17c783'}        2014              October   
4  {'$oid': '52b213b38594d8a2be17c784'}        2014              October   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   
2  2013-11-01T00:00:00Z   MINISTRY OF FINANCE AND ECONOMIC DEVEL   
3  2013-10-31T00:00:00Z   MIN. OF PLANNING AND INT'L COOPERATION   
4  2013-10-31T00:00:00Z                      MINISTRY OF FINANCE   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   
2                   NaN                                   Tuvalu!$!TV   
3                   NaN                        Republic of Yemen!$!RY   
4  2019-04-30T00:00:00Z                       Kingdom of Lesotho!$!LS   

  countrycode                              countryname    countryshortname  \
0          ET  Federal Democratic Republic of Ethiopia            Ethiopia   
1          TN                      Republic of Tunisia             Tunisia   
2          TV                                   Tuvalu              Tuvalu   
3          RY                        Republic of Yemen  Yemen, Republic of   
4          LS                       Kingdom of Lesotho             Lesotho   

                         ...                           sectorcode source  \
0                        ...                          ET,BS,ES,EP   IBRD   
1                        ...                                BZ,BS   IBRD   
2                        ...                                   TI   IBRD   
3                        ...                                   JB   IBRD   
4                        ...                             FH,YW,YZ   IBRD   

   status  supplementprojectflg  \
0  Active                     N   
1  Active                     N   
2  Active                     Y   
3  Active                     N   
4  Active                     N   

                                              theme1  \
0      {'Percent': 100, 'Name': 'Education for all'}   
1  {'Percent': 30, 'Name': 'Other economic manage...   
2    {'Percent': 46, 'Name': 'Regional integration'}   
3  {'Percent': 50, 'Name': 'Participation and civ...   
4  {'Percent': 30, 'Name': 'Export development an...   

                                      theme_namecode    themecode   totalamt  \
0      [{'code': '65', 'name': 'Education for all'}]           65  130000000   
1  [{'code': '24', 'name': 'Other economic manage...        54,24          0   
2  [{'code': '47', 'name': 'Regional integration'...  52,81,25,47    6060000   
3  [{'code': '57', 'name': 'Participation and civ...        59,57          0   
4  [{'code': '45', 'name': 'Export development an...        41,45   13100000   

  totalcommamt                                                url  
0    130000000  http://www.worldbank.org/projects/P129828/ethi...  
1      4700000  http://www.worldbank.org/projects/P144674?lang=en  
2      6060000  http://www.worldbank.org/projects/P145310?lang=en  
3      1500000  http://www.worldbank.org/projects/P144665?lang=en  
4     13100000  http://www.worldbank.org/projects/P144933/seco...  

[5 rows x 50 columns]

In [2]:
countries = wbp_df.countryname          #extracts only the countries column from the data 
countries.head()

0    Federal Democratic Republic of Ethiopia
1                        Republic of Tunisia
2                                     Tuvalu
3                          Republic of Yemen
4                         Kingdom of Lesotho
Name: countryname, dtype: object

In [3]:
top_countries = countries.value_counts()[:10]      #aggregates the countries by their counts, and slices the top 10 counts
top_countries

People's Republic of China         19
Republic of Indonesia              19
Socialist Republic of Vietnam      17
Republic of India                  16
Republic of Yemen                  13
People's Republic of Bangladesh    12
Nepal                              12
Kingdom of Morocco                 12
Africa                             11
Republic of Mozambique             11
Name: countryname, dtype: int64

## 2.) Finding the top 10 major project themes:

In [4]:
major_projects = wbp_df.mjtheme_namecode     #extracts only the major themes relevant column from the data
major_projects.head()

0    [{'code': '8', 'name': 'Human development'}, {...
1    [{'code': '1', 'name': 'Economic management'},...
2    [{'code': '5', 'name': 'Trade and integration'...
3    [{'code': '7', 'name': 'Social dev/gender/incl...
4    [{'code': '5', 'name': 'Trade and integration'...
Name: mjtheme_namecode, dtype: object

In [5]:
list_of_codes = []
list_of_themes = []                  #adds the name of each code and theme from each row to a respective empty list 
for row in major_projects:
    for dic in row:
        list_of_codes.append(dic['code'])
        list_of_themes.append(dic['name'])
top_themes = pd.Series(list_of_themes).value_counts()[:10]     #aggregates a total count for each theme in the list..    
top_themes                                                                 #then slices the top 10

Environment and natural resources management    223
Rural development                               202
Human development                               197
Public sector governance                        184
Social protection and risk management           158
Financial and private sector development        130
                                                122
Social dev/gender/inclusion                     119
Trade and integration                            72
Urban development                                47
dtype: int64

However... it looks like there are 122 missing themes with codes assigned to them. Let's fix that.

## 3.) Finding the top 10 major project themes (with missing values corrected):

In [6]:
code_themes_df = pd.DataFrame.from_items([('Code', list_of_codes), ('Theme', list_of_themes)])
code_themes_df.head()                                       #creates dataframe from the extracted codes and themes

Code                                  Theme
0    8                      Human development
1   11                                       
2    1                    Economic management
3    6  Social protection and risk management
4    5                  Trade and integration

In [7]:
import numpy as np

code_themes_df.replace(r'', np.nan, regex=True, inplace=True)                #replaces all empty themes with NaN
code_themes_df = code_themes_df.sort_values(by=['Code', 'Theme'])            #sorts the df by code and theme
code_themes_df.tail()                                               #prints tail of df to see if NaN are indeed present 

Code              Theme
1473    9  Urban development
1495    9  Urban development
333     9                NaN
460     9                NaN
471     9                NaN

In [8]:
code_themes_df['Theme'] = code_themes_df[['Theme']].fillna(method='ffill')  #forward fills the NaN values
code_themes_df.tail()                               #reprints tail of df to ensure NaN values were correctly filled

Code              Theme
1473    9  Urban development
1495    9  Urban development
333     9  Urban development
460     9  Urban development
471     9  Urban development

In [9]:
top_themes_filled = code_themes_df.Theme.value_counts()[:10]
top_themes_filled

Environment and natural resources management    250
Rural development                               216
Human development                               210
Public sector governance                        199
Social protection and risk management           168
Financial and private sector development        146
Social dev/gender/inclusion                     130
Trade and integration                            77
Urban development                                50
Economic management                              38
Name: Theme, dtype: int64

Now all 122 missing values are filled and accounted for.

## END